src/lib/integration_test_cloud_engine/kubernetes_network.ml

```diff
@@ -358,16 +359,18 @@ module Node = struct
     |> Deferred.bind ~f:Malleable_error.or_hard_error
 
   let dump_archive_data ~logger (t : t) ~data_file =
+    (* this function won't work if t doesn't happen to be an archive node *)
     let open Malleable_error.Let_syntax in
     [%log info] "Dumping archive data from (node: %s, container: %s)" t.pod_id
       t.container_id ;
-    let%map data =
-      Deferred.bind ~f:Malleable_error.return
-        (run_in_postgresql_container t
-           ~cmd:
-             "pg_dump --create --no-owner \
-              postgres://postgres:foobar@localhost:5432/archive")
+    let%bind dat =
+      (* (Deferred.bind ~f:Malleable_error.return *)
+      run_in_container t
+        ~cmd:
+          "pg_dump --create --no-owner \
+           postgres://postgres:foobar@archive-1-postgresql-0:5432/archive"
     in
+    let%map data = Deferred.bind ~f:Malleable_error.return dat in
```

Comment:
@nholland94 what is the correct way to write this, when you gotta do this double unwrap thing?

Reply:
I think the root issue here is that the return type for `run_in_container` doesn't really make sense. It doesn't have a failure mode (throws an exn if something goes wrong). The return value should just be `string Deferred.t`.

In general, we shouldn't use `Malleable_error.t` for functions that don't accumulate errors into the test context (except in some circumstances where it is more convenient to do so).  It's much easier to cast other monads (`Deferred.t`, `Or_error.t`, combinations of those) into a `Malleable_error.t` later than it is to extract information from a `Malleable_error.t`.